### PR TITLE
INT-3796: Configure SFTP Unknown Host Behavior

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3796

Previously, the `DefaultSftpSessionFactory` unconditionally accepted keys
from unknown hosts, or when a host key changed.

Permit this behavior to be configured with `allowUnknownHosts` and/or a custom
`UserInfo` properties in the factory.

`allowUknownHosts` is false by default (since 4.2) and true by default in 4.1.x for
backwards compatibility.

### CODE REVIEW

**Comment by garyrussell** [comment]
After review/polish, I will backport with boolean `true` and docs.

File: spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java [hunk 1]
```diff
@@ -43,11 +47,22 @@
...
  *
  * @since 2.0
  */
 public class DefaultSftpSessionFactory implements SessionFactory<LsEntry>, SharedSessionCapable {
 
+	private static final Log logger = LogFactory.getLog(DefaultSftpSessionFactory.class);
+
+	private final ReentrantReadWriteLock sharedSessionLock = new ReentrantReadWriteLock();
+
+	private final UserInfoWrapper userinfoWrapper = new UserInfoWrapper();
```
Comment (artembilan) [inline spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java]: Minor, but the code should be ideal :smile: 
Interfaces for properties and the name for the last one must be fixed as well - user**I**nfoWrapper

**Comment by garyrussell** [comment]
Pushed test fixes

File: spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java [hunk 2]
```diff
@@ -311,6 +324,41 @@ public void setEnableDaemonThread(Boolean enableDaemonThread){
...
+	 * @since 4.1.7
+	 */
+	public void setUserInfo(UserInfo userInfo) {
+		this.userInfo = userInfo;
+	}
+
+	/**
+	 * When no {@link UserInfo} has been provided, set to true to unconditionally allow
+	 * connecting to an unknown host or when a host's key has changed (see
+	 * {@link #setAllowUnknownKeys(boolean) unknownHosts}). Default false (since 4.2).
```
Comment (artembilan) [inline spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java]: It's just curious what you have had in mind:
javadocs

**Comment by artembilan** [comment]
Well, that's all for me here. 
That's all.
Just let me know about that my concern for JavaDocs and I can address all those minor issues on merge.
Seems for me the backport must be done only from the code perspective - without docs.

File: spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java [hunk 3]
```diff
@@ -428,50 +476,124 @@ public void resetSharedSession() {
...
-
-		private final String password;
+	private class UserInfoWrapper implements UserInfo, UIKeyboardInteractive {
+
+		/**
+		 * Convenience to check whether enclosing factory's UserInfo is configured.
+		 * @return true if there's a delegate.
+		 */
+		private boolean hasDelegate()
+		{
```
Comment (artembilan) [inline spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java]: It's isn't our code formatting.

**Comment by garyrussell** [comment]
Pushed. Correct (backport) except the default must be true.

**Comment by artembilan** [comment]
LGTM
Merging and backporting, considering the comments above.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/95a9e0d2dd88213828226bff4c90f399c6d8145a and backported as https://github.com/spring-projects/spring-integration/commit/7d559789ffa72f1a6d916cde81e2296b89d70b5a

**Pull request closed without merging**